features/79-structured-logging/implementation.md

```diff
@@ -0,0 +1,119 @@
+# Implementation
+
+### skygear-server
+
+- Log function to print structured logs
+- skygear-server create id per request and pass request id to plugin through context
+- Review all logs and add tag by its functional group (auth, push, pubsub ..etc)
+- Output JSON formatted logs on the skygear cloud. Output readable log when deployed in docker compose. Skygear Cloud can supply custom environment variable to enable this behavior.
+
+#### py-skygear, skygear-SDK-JS
+
+- Log function to print structured logs
+- Inject request id when printing structured logs
+- Review all logs and reduce redundant logs in all plugins, adding corresponding plugin tag
+- Output JSON formatted logs on the skygear cloud. Output readable log when deployed in docker compose. Skygear Cloud can supply custom environment variable to enable this behavior.
+
+### Portal
+
+- Filtering will be done in client side
+- Group logs by request ID option, those logs without request id will not be shown in this mode
+- With log level coloring
+- Increase logs size (Current: 1000)
+
+### skycli
+
+- Filtering will be done in client side, filters are passed by arguments
+- With log level coloring
+
+```sh
+skycli logs
+
+Options:
+  --level       Apply log level filter, comma separated. Support values: debug,info,warning,error,critical
+  --process     Apply process filter, comma separated. Support values: skygear-server,python,js
+  --tag         Apply tag filter, comma separated. Support values: request,auth,db,pubsub,push,auth_plugin,chat_plugin,cms_plugin,sso_plugin,others
+
+Examples:
+  # select skygear server logs and log level is debug
+  skycli --level=debug --process=skygear-server logs
+
+  # select cms and error logs
+  skycli --tags=cms,others logs
+
+  # select cloud functions logs
+  skycli --tags=others logs
+````
+
+### Cloud
+
+- Need to be able to retrieve logs from completed pod
+
+### Structured log format
+
+```js
+{
+    "time": 1521712900.511476,
```

Comment:
I think a ISO8601 timestamp would be more descriptive, something like `"2018-04-26T03:09:18.279977682Z"`. Oh if you have a good reference where you draw inspiration from, let me know 😀